File: dynamic-programming/720-longest-word-in-dictionary/longest-word-in-dictionary.cpp
class Solution {
public:
    struct cmp {
        bool operator()(const string &a, const string &b) {
            if (a.size() == b.size())
                return a < b;

            return a.size() > b.size();
        }
    };

    string longestWord(vector<string>& words) {

        unordered_set<string> st(words.begin(), words.end());

        sort(words.begin(), words.end(), cmp());

        string ans = "";

        for (auto &word : words) {

            bool valid = true;

            for (int i = 1; i < word.size(); i++) {

                if (!st.count(word.substr(0, i))) {
                    valid = false;
                    break;
                }
            }

            if (valid){
                ans = word;
                break;
            }
        }

        return ans;
    }
};

File: dynamic-programming/720-longest-word-in-dictionary/longest-word-in-dictionary.java
class Solution {
    public String longestWord(String[] words) {

        Set<String> wordSet = new HashSet<>();

        for (String word : words) {
            wordSet.add(word);
        }

        Arrays.sort(words, new Comparator<String>() {
            @Override
            public int compare(String a, String b) {

                if (b.length() != a.length()) {
                    return b.length() - a.length();
                }

                return a.compareTo(b);
            }
        });

        for (String word : words) {

            boolean good = true;

            for (int k = 1; k < word.length(); ++k) {

                if (!wordSet.contains(word.substring(0, k))) {
                    good = false;
                    break;
                }
            }

            if (good) {
                return word;
            }
        }

        return "";
    }
}